src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -0,0 +1,2957 @@
+  IL_0054:  br.s       IL_0057
+  IL_0056:  ldc.i4.0
+  IL_0057:  box        ""bool""
+  IL_005c:  call       ""string string.Format(string, object)""
+  IL_0061:  call       ""void System.Console.Write(string)""
+  IL_0066:  nop
+  IL_0067:  ret
+}");
+        }
+
+        [Fact]
+        public void TestNotEqualOnNullableVsNullableTuples()
+        {
+            var source = @"
+class C
+{
+    public static void Main()
+    {
+        Compare(null, null);
+        Compare(null, (1, 2));
+        Compare((2, 3), null);
+        Compare((4, 4), (4, 4));
+        Compare((5, 5), (10, 10));
+    }
+    private static void Compare((int, int)? nt1, (int, int)? nt2)
+    {
+        System.Console.Write($""{nt1 != nt2} "");
+    }
+}
+";
+            var comp = CreateStandardCompilation(source, references: s_valueTupleRefs, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "False True True False True");
+
+            var tree = comp.SyntaxTrees.First();
+            var model = comp.GetSemanticModel(tree);
+
+            var comparison = tree.GetCompilationUnitRoot().DescendantNodes().OfType<BinaryExpressionSyntax>().Single();
+            var nt1 = comparison.Left;
+            var nt1Type = model.GetTypeInfo(nt1);
+            Assert.Equal("nt1", nt1.ToString());
+            Assert.Equal("(System.Int32, System.Int32)?", nt1Type.Type.ToTestDisplayString());
+            Assert.Equal("(System.Int32, System.Int32)?", nt1Type.ConvertedType.ToTestDisplayString());
+
+            var nt2 = comparison.Right;
+            var nt2Type = model.GetTypeInfo(nt2);
+            Assert.Equal("nt2", nt2.ToString());
+            Assert.Equal("(System.Int32, System.Int32)?", nt2Type.Type.ToTestDisplayString());
+            Assert.Equal("(System.Int32, System.Int32)?", nt2Type.ConvertedType.ToTestDisplayString());
+
+            verifier.VerifyIL("C.Compare", @"{
+  // Code size      107 (0x6b)
+  .maxstack  3
+  .locals init ((int, int)? V_0,
+                bool V_1,
+                System.ValueTuple<int, int> V_2,
+                (int, int)? V_3,
+                System.ValueTuple<int, int> V_4)
+  IL_0000:  nop
+  IL_0001:  ldstr      ""{0} ""
+  IL_0006:  ldarg.0
+  IL_0007:  stloc.0
+  IL_0008:  ldloca.s   V_0
+  IL_000a:  call       ""bool (int, int)?.HasValue.get""
+  IL_000f:  stloc.1
+  IL_0010:  ldarg.1
+  IL_0011:  stloc.3
+  IL_0012:  ldloc.1
+  IL_0013:  ldloca.s   V_3
+  IL_0015:  call       ""bool (int, int)?.HasValue.get""
+  IL_001a:  beq.s      IL_001f
+  IL_001c:  ldc.i4.1
+  IL_001d:  br.s       IL_005a
+  IL_001f:  ldloc.1
+  IL_0020:  brtrue.s   IL_0025
+  IL_0022:  ldc.i4.0
+  IL_0023:  br.s       IL_005a
+  IL_0025:  ldloca.s   V_0
+  IL_0027:  call       ""(int, int) (int, int)?.GetValueOrDefault()""
+  IL_002c:  stloc.2
+  IL_002d:  ldloca.s   V_3
+  IL_002f:  call       ""(int, int) (int, int)?.GetValueOrDefault()""
+  IL_0034:  stloc.s    V_4
```

Comment:
Is there a test comparing long tuples, not just `Rest`?

Reply:
I'll add in next PR.
Took a note in [https://github.com/dotnet/roslyn/issues/22937](https://github.com/dotnet/roslyn/issues/22937)

---
In reply to: [170413123](https://github.com/dotnet/roslyn/pull/23356#discussion_r170413123) [](ancestors = 170413123)

Reply:
Updated `TestEvaluationOrderOnTupleLiteral`

---
In reply to: [170989353](https://github.com/dotnet/roslyn/pull/23356#discussion_r170989353) [](ancestors = 170989353,170413123)